docs/rfcs/0002-ecommerce.md

```diff
@@ -0,0 +1,176 @@
+the information.
+
+We didn't add any support for purchase orders to MicroMasters. I think we manually added information
+via Django shell to the `Order` model. We will need an admin site to allow admins to fill in purchase
+order information so that they don't manipulate these models directly.
+
+##### Recommendation
+
+If we roll our own we should use the Django admin for this. django-oscar probably has an interface
+for this already though I'm not familiar with it.
+
+#### Coupons
+
+edX overrides the django-oscar `Voucher` class to support coupons. They have several 
+different variations of coupons, for example:
+
+ - who can redeem: once by one user, once by many users, many times by many users
+ - discount: fixed discount or percent discount
+ - target: verified seat or audit in a course (or something else?)
+ - time of validity
+
+edX only allows one coupon to be used at checkout so they don't need to worry about order of operations
+when two coupons are applied. There is an admin site which allow admins to generate a CSV of coupon codes,
+and to create program offers, enterprise coupons and other kinds of offers. (I'm not sure how the others are used
+within edX.)
+
+In MicroMasters we had a custom implementation of coupons, working alongside a custom implementation of financial aid.
+The Django admin is used to create coupons and to record information about coupon invoices for bulk instances of
+coupons. 
+
+There are variations of coupons in MicroMasters:
+
+ - standard or discounted_previous_course: The latter handles a special case where people who already paid for
+ a previous course on edX get a discounted rate.
+ - discount: price discount, percent discount, or just set a new price.
+ - target: either program or course
+ - time of validity
+
+If a coupon is `standard`, the user "attaches" it by clicking on a link in an email. This directs
+them to a page and the code is taken off the URL and sent to the server. This creates a `UserCoupon`
+object to connect them to the coupon. After an order is created (before it is fulfilled) it
+creates a `RedeemedCoupon` to mark that the user is intending to make use of the coupon. When
+CyberSource posts back with the success message, this marks the `Order` as fulfilled and also
+marks the `Coupon` as having been used. The logic for figuring out if, when and where a coupon
+can be used is in `ecommerce.api`.
+ 
+##### Recommendation
+
+Whatever coupon system we use or come up with, we should assume one coupon per purchase and that
+every coupon has a coupon code. This greatly simplifies the logic and workflow. If a user
+has two coupons to choose from they would need to determine which is best for their particular
+purchase.
+
+We also may need to support support automatic discounts in the future
+based on arbitrary criteria, to be defined. It would be ideal to support this use case using
+coupons, given to the user and used explicitly like other coupons.
```

Comment:
I don't know of any use cases for "automatic" discounts.

Reply:
We can leave this out and come up with a way to support those use cases in the future using the existing coupons.